pkg/connection/store/kubernetes/store.go

```diff
@@ -0,0 +1,168 @@
+	}
+
+	// Configure client for an external API server with a given Kubeconfig.
+	kfg, err := resource.CommonCredentialExtractor(ctx, cfg.Kubernetes.Auth.Source, local, cfg.Kubernetes.Auth.CommonCredentialSelectors)
+	if err != nil {
+		return nil, errors.Wrap(err, errExtractKubernetesAuthCreds)
+	}
+	remote, err := clientForKubeconfig(kfg)
+	if err != nil {
+		return nil, errors.Wrap(err, errExtractKubernetesAuthCreds)
+	}
+
+	return &SecretStore{
+		client: remote,
+		applicator: resource.NewApplicatorWithRetry(resource.NewAPIPatchingApplicator(remote),
+			resource.IsAPIErrorWrapped, nil),
+		defaultNamespace: cfg.DefaultScope,
+	}, nil
+}
+
+// ReadKeyValues reads and returns key value pairs for a given Kubernetes Secret.
+func (ss *SecretStore) ReadKeyValues(ctx context.Context, i store.Secret) (store.KeyValues, error) {
+	s := &corev1.Secret{}
+	return s.Data, errors.Wrapf(ss.client.Get(ctx, types.NamespacedName{Name: i.Name, Namespace: ss.namespaceForSecret(i)}, s), errGetSecret)
+}
+
+// WriteKeyValues writes key value pairs to a given Kubernetes Secret.
+func (ss *SecretStore) WriteKeyValues(ctx context.Context, i store.Secret, kv store.KeyValues) error {
+	meta := secretMetadata{}
+	if len(i.Metadata) > 0 {
+		if err := json.Unmarshal(i.Metadata, &meta); err != nil {
+			return errors.Wrap(err, errParseMetadata)
+		}
+	}
+
+	t := resource.SecretTypeConnection
+	if meta.Type != "" {
+		t = meta.Type
+	}
+
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        i.Name,
+			Namespace:   ss.namespaceForSecret(i),
+			Labels:      meta.Labels,
+			Annotations: meta.Annotations,
+		},
+		Type: t,
+		Data: kv,
+	}
+
+	return errors.Wrap(ss.applicator.Apply(ctx, s), errCreateOrUpdateSecret)
+}
+
+// DeleteKeyValues delete key value pairs from a given Kubernetes Secret.
+// If no kv specified, the whole secret instance is deleted.
+// If kv specified, those would be deleted and secret instance will be deleted
+// only if there is no data left.
+func (ss *SecretStore) DeleteKeyValues(ctx context.Context, i store.Secret, kv store.KeyValues) error {
+	// NOTE(turkenh): DeleteKeyValues method wouldn't need to do anything if we
+	// have used owner references similar to existing implementation. However,
+	// this wouldn't work if the K8s API is not the same as where SecretOwner
+	// object lives, i.e. a remote cluster.
+	// Considering there is not much additional value with deletion via garbage
+	// collection in this specific case other than one less API call during
+	// deletion, I opted for unifying both instead of adding conditional logic
+	// like add owner references if not remote and not call delete etc.
+	s := &corev1.Secret{}
+	err := ss.client.Get(ctx, types.NamespacedName{Name: i.Name, Namespace: ss.namespaceForSecret(i)}, s)
+	if kerrors.IsNotFound(err) {
+		// Secret already deleted, nothing to do.
+		return nil
+	}
+	if err != nil {
+		return errors.Wrap(err, errGetSecret)
+	}
+	// Delete all keys from secret data
```

Comment:
Nit: Maybe "delete all supplied keys"? At first I read this as if we always deleted all keys from the secret.